Mysensors nodes can be renamed in config file

EmitKiwi opened the pull request:
## Description:

Make it possible to change the node names. Node configuration is now independent from the node id. Also replacing a node is easier.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3236

## Example entry for `configuration.yaml` (if applicable):
```yaml
mysensors:
  gateways:
  - device: '/dev/ttyUSB0'
    baud_rate: 115200
    nodes:
      1:
        name: 'kitchen'
      3: 
        name: 'living_room'
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

EmitKiwi committed:
Mysensors nodes can be renamed in the config file

homeassistant commented:
Hi @EmitKiwi, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#9123).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented:
I'm hesitant about allowing the user to configure the entity_id. It's already possible to change the friendly_name in customize config section. It's also possible for the user to influence the entity_id by selecting a sketch name, node_id and child_id in the sketch.

Why should we allow the entity_id to be changed? How is it easier to replace a node?

EmitKiwi commented:
I made this pull request, because I do not want to set my Node ID manually in the sketch. It is specific for people who want to use the automatically distributed node ids from the gateway. If you want to replace a node in this situation (or when you reset the gateway), you would need to edit all you're configuration (automation) files where that specific node is mentioned. In this pull request, the only thing you need to do is to change the id's in the configuration.

MartinHjelmare commented:
Good motivation!

MartinHjelmare commented on homeassistant/components/mysensors.py:
```diff
@@ -120,6 +124,14 @@ def validator(config):
+NODE_SCHEMA = vol.All(
+    cv.ensure_list,
+    [{
+        vol.Required(CONF_NODE_ID): cv.positive_int,
```
Remove the ID key. The number key on the dict will be the ID.

MartinHjelmare commented on homeassistant/components/mysensors.py:
```diff
@@ -120,6 +124,14 @@ def validator(config):
+    cv.ensure_list,
+    [{
+        vol.Required(CONF_NODE_ID): cv.positive_int,
+        vol.Optional(CONF_NODE_NAME): cv.string
```
Name should be required in this part of the schema.

MartinHjelmare commented on homeassistant/components/mysensors.py:
```diff
@@ -139,6 +151,7 @@ def validator(config):
                     CONF_TOPIC_IN_PREFIX, default=''): valid_subscribe_topic,
                 vol.Optional(
                     CONF_TOPIC_OUT_PREFIX, default=''): valid_publish_topic,
+                vol.Optional(CONF_NODES, default=[]): NODE_SCHEMA,
```
Default to empty dict.

MartinHjelmare commented on homeassistant/components/mysensors.py:
```diff
@@ -286,8 +300,14 @@ def mysensors_callback(gateway, msg):
+            node_name = msg.node_id
+            for _, node in enumerate(gateway.nodes_config):
+                if (node.get(CONF_NODE_NAME) is not None and
+                        node.get(CONF_NODE_ID) is msg.node_id):
```
Compare with `==` not `is`.

MartinHjelmare commented on homeassistant/components/mysensors.py:
```diff
@@ -286,8 +300,14 @@ def mysensors_callback(gateway, msg):
+                    break
             name = '{} {} {}'.format(
-                gateway.sensors[msg.node_id].sketch_name, msg.node_id,
+                gateway.sensors[msg.node_id].sketch_name, node_name,
```
I think the whole name should be replaced, not just the node id. If the user wants to change name, the user takes responsibility for the whole name.

MartinHjelmare reviewed (changes requested):
You'll need to update the iteration based on a dict instead of a list.

FYI, there's a big mysensors refactor PR open now too, but if we merge this first I can make it work in that PR. But we'll have to see what's get merged first.

EmitKiwi commented on homeassistant/components/mysensors.py:
```diff
@@ -120,6 +124,14 @@ def validator(config):
 
 
+NODE_SCHEMA = vol.All(
+    cv.ensure_list,
```
How do I validate a dict in a dict, without knowing the dict key?

MartinHjelmare replied:
See eg here:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/climate/heatmiser.py#L31-L32

EmitKiwi committed:
Replace nodes array with dict. Replace whole name of the node.

EmitKiwi commented:
I changed the list to the dict object and made it possible to rename the whole name. (Sketch name + Node id).

MartinHjelmare commented on homeassistant/components/mysensors.py:
```diff
@@ -300,15 +297,14 @@ def mysensors_callback(gateway, msg):
-            for _, node in enumerate(gateway.nodes_config):
+            node_name = '{} {}'.format(
+                gateway.sensors[msg.node_id].sketch_name, msg.node_id)
+            for node_id, node in gateway.nodes_config.items():
```
```py
node_name = next(
    (node[CONF_NODE_NAME] for node_id, node in gateway.nodes_config.items()
    if node.get(CONF_NODE_NAME) is not None and node_id == msg.node_id),
    node_name)
```

MartinHjelmare reviewed (changes requested):
Looks good! One code enhancement that can be done.

MartinHjelmare commented:
https://github.com/home-assistant/home-assistant/pull/9069 has been merged, which means we have to adjust the approach in this PR. First rebase on dev branch. Then set the nodes_config directly as an attribute on the gateway after it's been created in `setup_gateway`, and modify `get_mysensors_name` to return the correct name.

EmitKiwi committed:
Improved iteration on node names

homeassistant commented:
Hello @EmitKiwi,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`0c351851049c071740e1981a74108ae9661190ec`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;9123&#x2F;commits&#x2F;0c351851049c071740e1981a74108ae9661190ec) - This commit has something that looks like an email address (gitlab@jiproos.nl). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

EmitKiwi commented:
Fixed the cla-error (commited with the wrong email). Is it possible to delete that comment?

I rebased as requested and did the fixes.

MartinHjelmare reviewed (changes requested):
Code looks good! Please update the configuration example in the description and this is good to go.

EmitKiwi commented:
I updated the configuration example and added the description of the structure from the entity_id in the [documentation](https://github.com/home-assistant/home-assistant.github.io/pull/3236).

MartinHjelmare reviewed (approved):
Great!

EmitKiwi commented:
Thanks for the great support!
